momepy/coins.py

```diff
@@ -0,0 +1,572 @@
+    angle_threshold : int, float (default 0)
+        the angle threshold for COINS algorithm. Segments will only be considered 
+        a part of the same street if deflection angle is above the threshold.
+    out_gdf : string (default "stroke_attribute")
+        the desired output GeoDataFrame. 
+        -"stroke_attribute" returns the original GeoDataFrame with stroke group 
+          as a new attribute.
+        -"strokes" returns a new GeoDataFrame with strokes as the edges.
+          Refer to COINS paper for more details.
+        -"premerge" returns a new GeoDataFrame prior to performing merging
+          algorithm. Refer to COINS paper for more details
+
+    Returns
+    ----------
+    edge_gdf_new : GeoDataFrame
+        output GeoDataFrame based on "out_gdf"
+
+    Examples
+    --------
+    >>> strokes_gdf = momepy.COINS(edge_gdf, out_gdf="strokes")
+    
+    """
+
+    # Start time
+    t1 = time.time()
+
+    # Process edge geodataframe
+    myStreet = line(edge_gdf)
+
+    # Split edges into individual line segements
+    myStreet.splitLines()
+
+    # Create unique id for each indivdual line segment
+    myStreet.uniqueID()
+
+    # Compute edge connectivity table
+    myStreet.getLinks()
+
+    # Find best link at every point for both lines
+    myStreet.bestLink()
+
+    # Cross check best links and enter angle threshold for connectivity
+    myStreet.crossCheckLinks(angle_threshold)
+
+    if out_gdf == 'premerge':
+
+        # Export pre merged edges
+        edge_gdf_new = myStreet.create_gdf_premerge()
+
+    elif out_gdf == 'strokes':
+
+        # Merge lines into strokes
+        myStreet.mergeLines()
+
+        # Export edges with strokes
+        edge_gdf_new = myStreet.create_gdf_strokes()
+
+    elif out_gdf == 'stroke_attribute':
+        
+        # Merge lines into strokes
+        myStreet.mergeLines()
+
+        # Add stroke attribute to original dataframe
+        edge_gdf_new = myStreet.add_gdf_stroke_attributes()
+    
+    # End time
+    t2 = time.time()
+    
+    minutes = math.floor((t2-t1) / 60)
+    seconds = (t2 - t1) % 60
+    print("Processing complete in %d minutes %.2f seconds." % (minutes, seconds))
+
+    return edge_gdf_new
+
+class line:
+
+    def __init__(self, edge_gdf):
+        self.edge_gdf = edge_gdf
+        self.gdfProjection = self.edge_gdf.crs
+
+        # Get indices of original gdf
+        self.uv_index =  self.edge_gdf.index.tolist()
+
+        # Get line segments from edge gdf
+        self.lines = [list(value[1].coords) for value in edge_gdf['geometry'].iteritems()]
+
+    def splitLines(self):
+        outLine = []
+        tempLine = []
+        self.tempArray = []
+        n = 0
+        #Iterate through the lines and split the edges
+        idx = 0
+        for line in self.lines:
+            for part in _listToPairs(line):
+                outLine.append([part, _computeOrientation(part), list(), list(), list(), list(), list(), list(), self.uv_index[idx]])
```

Comment:
What are these empty lists for?

Reply:
These lists are used in the negotiation to choose the stroke group.  If you look at `create_gdf_premerge()` you can see what each list represents.

- linksP1
- linksP2
- bestP1
- bestP2
- P1Final
- P2Final

These empty lists get filled in the `getLinks()`, `bestLink()` and `crossCheckLinks()` methods.

You can refer to Table 2 from the [paper](https://journals.sagepub.com/doi/10.1177/2399808320967680) to see what each variable represents.

If you compare to the original [code](https://github.com/PratyushTripathy/COINS/blob/master/PythonTool/200330_NetworkContinuityMultiprocessing_PT_Rework_Final.py) (line 277), you will see that I added the last item of outLine. This is so that I can track the original edge that each segment belongs to.